### PR TITLE
(PE-11531) Allow upgrades from Puppet 4+

### PR DESCRIPTION
**Pull request by highb**
Prior to this commit we only allowed upgrades with this
module if the current puppet version was pre-4.
This commit removes that restriction and hopefully handles
all the edge cases involved in upgrade between new versions
and doesn't break FOSS. This feature should be tested
thoroughly.

### CODE REVIEW

**Comment by highb** [comment]
I've tested this out with 3.8->2015.3.2 as well as 2015.2.1->2015.3.2 upgrades on centos7 and windows 2012 and it appears that `agent_upgrade.rb` passes. We'll definitely need to test more platforms.

**Comment by jpartlow** [comment]
@highb tests are failing lint of manifests/osfamily/debian.pp - WARNING: variable not enclosed in {} on line 75.  I tried to push a commit fix to your branch but hit a permission issue (?)

Once that's fixed I'm seeing 70 spec errors around puppet_agent::service.  Are you not seeing those locally?

**Comment by highb** [comment]
@jpartlow Hm, nope I don't think I was seeing those spec failures locally. Maybe something is off in my bundler environment, I'll poke around after I've had a change to check out these other PRs on the board.

**Comment by highb** [comment]
@jpartlow OK, yeah was able to repro the spec failures locally. I'm not sure how I didn't see those before.  I'll get those fixed up.

**Comment by jpartlow** [comment]
@highb Now we're seeing spec failures just for Puppet 3.7.5 and Puppet 4.0

**Comment by highb** [comment]
@jpartlow Yep, I'm going to look into those once I've finished reviewing PRs.

File: manifests/init.pp [hunk 1]
```diff
@@ -66,16 +87,19 @@
...
+      package_version   => $package_version,
     }
 
     contain '::puppet_agent::prepare'
     contain '::puppet_agent::install'
 
     # On windows, our MSI handles the services
-    if $::osfamily != 'windows' {
+    # On PE AIO nodes, PE Agent nodegroup is managing the services
+    if $::osfamily != 'windows' and versioncmp($::clientversion, '4.0.0') < 0 {
```
Comment (MikaelSmith) [inline manifests/init.pp]: Shouldn't this be restricted to is_pe as well then?

Comment (highb) [inline manifests/init.pp]: Oh yes, that'd be good to check.

File: manifests/init.pp [hunk 2]
```diff
@@ -1,54 +1,75 @@
...
+    $_package_version = $package_version
   }
-  elsif versioncmp("${::clientversion}", '4.0.0') >= 0 {
-    info('puppet_agent performs no actions on Puppet 4+')
+
+  if versioncmp("${::clientversion}", '3.8.0') < 0 {
+    fail('upgrading requires at least Puppet 3.8')
   }
-  else {
+  elsif $is_pe and $::aio_agent_version != undef and $_package_version != undef and versioncmp("${::aio_agent_version}", "${_package_version}") >= 0 {
```
Comment (thrnio) [inline manifests/init.pp]: Why does this do nothing if up-to-date only if running on PE?

File: manifests/init.pp [hunk 3]
```diff
@@ -66,16 +87,19 @@
...
     class { '::puppet_agent::prepare':
       package_file_name => $_package_file_name,
+      package_version   => $package_version,
     } ->
     class { '::puppet_agent::install':
       package_file_name => $_package_file_name,
+      package_version   => $package_version,
     }
 
     contain '::puppet_agent::prepare'
```
Comment (MikaelSmith) [inline manifests/init.pp]: I think MODULES-3118 implies this needs to happen even if the aio_agent_version and package_version match.

File: manifests/init.pp (hunk 2 above)
Comment (MikaelSmith) [inline manifests/init.pp]: We should probably still manage all the resources, and handle ensuring we don't make changes when the package installed is the correct version on a more granular level.

Comment (highb) [inline manifests/init.pp]: That makes sense to me. I think the only `is_pe` checks we should be doing are for the service check (since PE already manages the services) and for the PE only platforms (AIX, etc)

Comment (thrnio) [inline manifests/init.pp]: :+1:

**Comment by highb** [comment]
@MikaelSmith Hm... we should probably think up a method for preventing that.

**Comment by MikaelSmith** [comment]
Or support it. I'm ok with simply documenting it for now.

**Comment by highb** [comment]
@MikaelSmith @thrnio I've finished up fixing the specs to test Puppet 4+, and addressed the lack of resource management in already upgraded PE nodes. I'm going to run some tests to verify PE doesn't break in some new and surprising way.

**Comment by highb** [comment]
OK, one more spec fix is needed for pre-4 behavior.

File: Gemfile [hunk 4]
```diff
@@ -2,7 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 
 group :test do
   gem 'rake', '~> 10.4'
-  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 3.8'
+  gem 'puppet', ENV['PUPPET_VERSION'] || '~> 4.0'
```
Comment (MikaelSmith) [inline Gemfile]: Why is this change needed?

Comment (highb) [inline Gemfile]: Oh, it isn't really. That's just what I've been using for testing. All the tests should be specifying PUPPET_VERSION, though so it seems like a sensible default to me?

Comment (MikaelSmith) [inline Gemfile]: Ok.

File: manifests/init.pp [hunk 5]
```diff
@@ -1,54 +1,72 @@
 # == Class: puppet_agent
 #
-# Upgrades Puppet 3.8 to Puppet 4+ (Puppet-Agent from Puppet Collection 1).
-# Makes the upgrade easier by migrating SSL certs and config files to the new
-# Puppet-Agent paths and removing deprecated settings that are no longer
+# Upgrades Puppet 3.8 and newer to the requested version.
```
Comment (MikaelSmith) [inline manifests/init.pp]: Worth being clear that it will only upgrade to a puppet-agent package.

File: manifests/init.pp [hunk 6]
```diff
@@ -1,54 +1,72 @@
...
   validate_re($arch, ['^x86$','^x64$','^i386$','^i86pc$','^amd64$','^x86_64$','^power$','^sun4[uv]$','PowerPC_POWER'])
 
-  if versioncmp("${::clientversion}", '3.8.0') < 0 {
-    fail('upgrading requires Puppet 3.8')
-  }
-  elsif versioncmp("${::clientversion}", '4.0.0') >= 0 {
-    info('puppet_agent performs no actions on Puppet 4+')
+  if ($package_version == undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } elsif ($package_version != undef and versioncmp("${::clientversion}", $package_version) != 0) {
```
Comment (MikaelSmith) [inline manifests/init.pp]: I'm not clear what this check is doing?

Also, is there an attempt to skip installing if the correct version is already installed?

Comment (MikaelSmith) [inline manifests/init.pp]: Can we copy the `aio_agent_version` fact to this module, so OSS upgrades can use it? We should be checking whether `aio_agent_version` matches requested version, and not performing an install on platforms without package managers (or maybe just skip it on all platforms).

Comment (thrnio) [inline manifests/init.pp]: It's kind of silly IMO that the `aio_agent_version` fact is only shipped with PE.

Comment (MikaelSmith) [inline manifests/init.pp]: This may not be a major problem... rerunning the MSI on Windows takes about 5 seconds.

Comment (MikaelSmith) [inline manifests/init.pp]: Filed https://tickets.puppetlabs.com/browse/PA-241 to add it.

Comment (MikaelSmith) [inline manifests/init.pp]: Something like
```
diff --git a/manifests/install.pp b/manifests/install.pp
index 60cdde2..806f219 100644
--- a/manifests/install.pp
+++ b/manifests/install.pp
@@ -16,6 +16,7 @@ class puppet_agent::install(
 ) {
   assert_private()

+  if ($::aio_agent_version != $package_version) {
     if ($::operatingsystem == 'SLES' and $::operatingsystemmajrelease == '10') or ($::operatingsystem == 'AIX' and  $::architecture =~ /PowerPC_POWER[5,6,7]/) {
       contain puppet_agent::install::remove_packages

@@ -78,3 +79,4 @@ class puppet_agent::install(
       }
     }
   }
+}
```
would work. We could look at adding a custom fact, or just plan to address it for OSS in a future puppet-agent release.

File: manifests/init.pp [hunk 7]
```diff
@@ -1,54 +1,72 @@
...
     } elsif $::operatingsystem == 'aix' and $::architecture =~ /PowerPC_POWER[5,6,7]/ {
       $aix_ver_number = regsubst($::platform_tag,'aix-(\d+\.\d+)-power','\1')
-      $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.aix${aix_ver_number}.ppc.rpm"
+      $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.aix${aix_ver_number}.ppc.rpm"
     } elsif $::operatingsystem == 'Darwin' and $::macosx_productversion_major =~ /10\.[9,10,11]/ {
-      $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.osx${$::macosx_productversion_major}.dmg"
+      $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.osx${$::macosx_productversion_major}.dmg"
     } elsif $::osfamily == 'windows' {
       $_arch = $::kernelmajversion ?{
         /^5\.\d+/ => 'x86', # x64 is never allowed on windows 2003
```
Comment (MikaelSmith) [inline manifests/init.pp]: This section needs to be updated to use the package version.

File: manifests/init.pp (hunk 6 above)
Comment (highb) [inline manifests/init.pp]: Hm, I guess we should just set `_package_version` if `package_version` is set. The version mismatch check was from when I was thinking the module would do a no-op if an upgrade wasn't needed.

File: manifests/init.pp (hunk 7 above)
Comment (highb) [inline manifests/init.pp]: I recall going through a similar line of thought where I assumed that there would be a version on the package, but I think the Windows MSIs don't have versions specified on the package name?

Comment (MikaelSmith) [inline manifests/init.pp]: Might depend on where we get them. OSS versions do: https://downloads.puppetlabs.com/windows/

Comment (highb) [inline manifests/init.pp]: Interesting. Currently the windows prepare step is very PE-centric. I think it depends on pe_repo being setup on the master. We could just document that if you want to use this module with FOSS, you should re-name your MSI?

Comment (jpartlow) [inline manifests/init.pp]: There's something odd about that, because I thought windows upgrades were working in FOSS, and what we did for PE, was just deal with a) we needed to limit to installing windows packages provided with that pe version via pm.puppetlabs.com (downloaded locally to pe_repo), b) the fact that the master's cert was not trusted by windows boxes as far as transferring the windows package.

Comment (MikaelSmith) [inline manifests/init.pp]: Upgrades work, but they always get the latest release. With the change to specify a specific version, that logic needs to be updated.

Comment (highb) [inline manifests/init.pp]: Ah, right.

File: manifests/init.pp (hunk 6 above)
Comment (MikaelSmith) [inline manifests/init.pp]: This should only be needed for package providers that don't support versionable (Mac and Solaris 10), and Windows because we do an out-of-band upgrade. It looks like every provider we'll encounter that supports versionable also supports upgradeable, so no extra concerns there.

File: manifests/install.pp [hunk 8]
```diff
@@ -61,9 +64,16 @@
         source            => $::puppet_agent::source,
       }
     }
+  } elsif ($::osfamily == 'Solaris' and $::operatingsystemmajrelease == '10') or $::osfamily == 'Darwin' {
+    # Solaris 10/OSX package provider does not provide 'versionable'
+    # Package is removed above, then re-added as the new version here.
+    package { $::puppet_agent::package_name:
+      ensure => 'present',
```
Comment (MikaelSmith) [inline manifests/install.pp]: How does 'present' ensure the correct package? Because we supply a source?

Comment (highb) [inline manifests/install.pp]: Yeah, since Solaris doesn't support versioned, we can't give a version here. So we just ensure that the source is the correct version.

**Comment by highb** [comment]
@MikaelSmith Alright, I think I've addressed all your feedback and rebased the PR since it had some new merge conflicts.

File: manifests/init.pp [hunk 9]
```diff
@@ -1,51 +1,72 @@
...
+  $source          = $::puppet_agent::params::_source,
 ) inherits ::puppet_agent::params {
 
   validate_re($arch, ['^x86$','^x64$','^i386$','^i86pc$','^amd64$','^x86_64$','^power$','^sun4[uv]$','PowerPC_POWER'])
 
-  if versioncmp("${::clientversion}", '4.0.0') >= 0 {
-    info('puppet_agent performs no actions on Puppet 4+')
+  if ($package_version == undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } else
```
Comment (MikaelSmith) [inline manifests/init.pp]: Missing an open curly brace, caused CI to fail.

File: manifests/install.pp [hunk 10]
```diff
@@ -7,9 +7,12 @@
 # [package_file_name]
 #   The puppet-agent package file name.
 #   (see puppet_agent::prepare::package_file_name)
+# [version]
+#   The puppet-agent version to install.
 #
 class puppet_agent::install(
   $package_file_name = undef,
+  $package_version   = 'present',
```
Comment (MikaelSmith) [inline manifests/install.pp]: Will this ever be used? We always pass an argument in init.pp, so seems like it will either be a version or undef.

File: manifests/init.pp (hunk 6 above)
Comment (MikaelSmith) [inline manifests/init.pp]: I didn't see any changes corresponding to this problem (not repeating installs on Mac, Solaris 10, and Windows).

File: manifests/init.pp (hunk 9 above)
Comment (highb) [inline manifests/init.pp]: Odd, I remember fixing that.

**Comment by highb** [comment]
I'll finish up the fixes for the specs and another rebase tomorrow.

**Comment by highb** [comment]
Rebased again, now going to work on the platforms that need their packages removed before upgrading.

File: manifests/init.pp [hunk 11]
```diff
@@ -1,46 +1,67 @@
...
       } else {
-        $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.i386.pkg.gz"
+        $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.i386.pkg.gz"
       }
     } elsif $::operatingsystem == 'Darwin' and $::macosx_productversion_major =~ /10\.[9,10,11]/ {
-      $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.osx${$::macosx_productversion_major}.dmg"
+      $_package_file_name = "${puppet_agent::package_name}-${_package_version}-1.osx${$::macosx_productversion_major}.dmg"
     } elsif $::operatingsystem == 'aix' and $::architecture =~ /PowerPC_POWER[5,6,7]/ {
       $aix_ver_number = regsubst($::platform_tag,'aix-(\d+\.\d+)-power','\1')
       $_package_file_name = "${puppet_agent::package_name}-${puppet_agent::params::master_agent_version}-1.aix${aix_ver_number}.ppc.rpm"
```
Comment (highb) [inline manifests/init.pp]: Hm, this `${puppet_agent::params::master_agent_version}` should be `${package_version}`.

Comment (MikaelSmith) [inline manifests/init.pp]: Looks like it.

**Comment by highb** [comment]
Specs were failing due to the linter not liking some double quoted strings w/o variables in them that I added. Fixed up. I've run the specs with Puppet 3.8 locally, hopefully nothing broke on 4.0

File: manifests/install/remove_packages.pp [hunk 12]
```diff
@@ -69,14 +67,22 @@
           'pe-ruby-ldap',
         ]
       }
+    } elsif versioncmp("${::aio_agent_version}", "${::package_version}") < 0 {
+      if $::operatingsystem == 'Solaris' {
+        $packages = [
+          'PUPpuppet-agent',
```
Comment (MikaelSmith) [inline manifests/install/remove_packages.pp]: This is necessary?

**Comment by MikaelSmith** [comment]
Looks good, except I think Windows will still unnecessarily re-install the MSI. You should guard windows/install.pp if aio_agent_version matches clientversion.

File: manifests/install/remove_packages.pp (hunk 12 above)
Comment (highb) [inline manifests/install/remove_packages.pp]: I think so, solaris is weird. I'll test it out to be sure.

Also, I still need to add SLES/AIX below to get around their weird, non-versionable package managers.

**Comment by highb** [comment]
@MikaelSmith Then it will only be a problem on FOSS until we ship `aio_agent_version` with facter, right?

**Comment by highb** [comment]
I'm testing all this out as I go along with pe_acceptance_tests/agent_upgrade.rb. I just added a minor change to agent_upgrade to add the package_version if we're upgrading from a PE_VER > 3.8.

File: manifests/install/remove_packages.pp (hunk 12 above)
Comment (highb) [inline manifests/install/remove_packages.pp]: Oh no, this is going to cause a duplicate resource declaration...

Comment (MikaelSmith) [inline manifests/install/remove_packages.pp]: Why?

Comment (highb) [inline manifests/install/remove_packages.pp]: Ah, just ran a solaris upgrade and it *is* just `puppet-agent` now! Fixing...

**Comment by highb** [comment]
Bah, didn't mean to push that up yet. Specs will fail because I haven't fixed some syntax issues and the specs still need another update. Update on testing: I've been working around the yum ensure => version issue that has been causing grief. It requires the `-1.el$majorversion` bit on the version, also just running 3.8.4/2015.3 tests takes quite a while...

**Comment by highb** [comment]
Most recent changes have been tested on 3.8.4->2016.1 upgrade for centos7, solaris10 and sles10. Going to try out 2015.3.3->2016.1 with the same platforms plus OSX1010.

File: .fixtures.yml [hunk 13]
```diff
@@ -1,6 +1,7 @@
 fixtures:
   repositories:
     stdlib: "https://github.com/puppetlabs/puppetlabs-stdlib.git"
+    transition: "https://github.com/puppetlabs/puppetlabs-transition.git"
```
Comment (highb) [inline .fixtures.yml]: Needed because otherwise we would have duplicate resource declarations on platforms that need to do the package remove/install dance to upgrade (Solaris/SLES/AIX)

File: manifests/install.pp [hunk 14]
```diff
@@ -50,20 +53,37 @@
...
-      class { 'puppet_agent::windows::install':
-        package_file_name => $package_file_name,
-        source            => windows_native_path("${::puppet_agent::params::local_packages_dir}/${package_file_name}"),
-      }
-    } else {
-      class { 'puppet_agent::windows::install':
-        package_file_name => $package_file_name,
-        source            => $::puppet_agent::source,
+    # Prevent re-running the batch install
+    if versioncmp("${::agent_version}", "${package_version}") < 0 {
```
Comment (MikaelSmith) [inline manifests/install.pp]: So this blocks downgrades. That seems ok for now, since we don't officially support them.

**Comment by MikaelSmith** [comment]
Failed CI. I'm :+1: on this in general though.

File: manifests/install/remove_packages.pp [hunk 15]
```diff
@@ -69,13 +73,24 @@
...
+      if (versioncmp("${::clientversion}", '4.0.0') < 0) {
         package { $old_package:
-          ensure => absent,
-          *      => $package_options,
+          * => $package_options,
+        }
+      } else {
+        # We must use transition here because we would have a duplicate package
+        # declaration if we used a Package.
+        notify { "using puppetlabs-transition to remove ${old_package}: ${::operatingsystem} does not support versionable": }
```
Comment (MikaelSmith) [inline manifests/install/remove_packages.pp]: You could use a name and message here to make the notify simpler.

Comment (MikaelSmith) [inline manifests/install/remove_packages.pp]: And don't you need to declare `package { $old_Package }` still?

Comment (highb) [inline manifests/install/remove_packages.pp]: Yeah, if I did declare it, that's where the duplicate resource errors come up.

Comment (highb) [inline manifests/install/remove_packages.pp]: Naming it would probably be a good call. :P

**Comment by highb** [comment]
@MikaelSmith Yeah, because I've added a module the Puppetfile isn't updated, so I had to hack around in https://github.com/puppetlabs/pe_acceptance_tests/blob/2016.2.x/setup/agent_upgrade.rb#L112 to add the transition module. :/

**Comment by highb** [comment]
Damn it, now the dev build version format is causing problems again. I guess I'll just make dev builds acceptable to the `package_version` regex.

File: manifests/init.pp [hunk 16]
```diff
@@ -1,49 +1,72 @@
...
-  if versioncmp("${::clientversion}", '4.0.0') >= 0 {
-    info('puppet_agent performs no actions on Puppet 4+')
+  if ($package_version == undef and $puppet_agent::params::master_agent_version != undef and versioncmp("${::clientversion}", '4.0.0') < 0) {
+    $_package_version = $puppet_agent::params::master_agent_version
+  } elsif $package_version != undef {
+    $_package_version = $package_version
   }
-  else {
+
+  if versioncmp("${::clientversion}", '3.8.0') < 0 {
```
Comment (MikaelSmith) [inline manifests/init.pp]: This didn't rebase correctly.

File: spec/classes/puppet_agent_osfamily_aix_spec.rb [hunk 17]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper'
 
 
-describe 'puppet_agent', :unless => Puppet.version < "3.8.0" || Puppet.version >= "4.0.0" do
+describe 'puppet_agent', :unless => Puppet.version < "3.8.0" do
```
Comment (MikaelSmith) [inline spec/classes/puppet_agent_osfamily_aix_spec.rb]: We can stop guarding against version less than 3.8. Missed that in the commit that dropped the 3.8 guard.

File: spec/classes/puppet_agent_spec.rb [hunk 18]
```diff
@@ -29,31 +34,58 @@
...
 
             it { is_expected.to compile.with_all_deps }
+            it { expect { catalogue }.not_to raise_error }
+          end
+        end
 
-            it { is_expected.to contain_class('puppet_agent') }
-            it { is_expected.to contain_class('puppet_agent::params') }
-            if Puppet.version < "4.0.0"
+        if Puppet.version < "3.8.0"
```
Comment (MikaelSmith) [inline spec/classes/puppet_agent_spec.rb]: This rebased incorrectly.

File: manifests/install.pp [hunk 19]
```diff
@@ -50,20 +53,36 @@
...
+    # Solaris 10/OSX/AIX/SLES 10 package provider does not provide 'versionable'
+    # Package is removed above, then re-added as the new version here.
+    package { $::puppet_agent::package_name:
+      ensure => 'present',
+      *      => $_package_options,
+    }
+  } elsif ($::osfamily == 'RedHat') and ($package_version != 'present') {
+    # Workaround PUP-5802/PUP-5025
+    package { $::puppet_agent::package_name:
+      ensure => "${package_version}-1.el${::operatingsystemmajrelease}",
```
Comment (MikaelSmith) [inline manifests/install.pp]: Yes, operatingsystemmajrelease is the right way to do this.

File: spec/classes/puppet_agent_spec.rb (hunk 18 above)
Comment (highb) [inline spec/classes/puppet_agent_spec.rb]: D'oh. Good catch.

File: manifests/install.pp (hunk 19 above)
Comment (thrnio) [inline manifests/install.pp]: I thought 3.8 and stringified facts disabled was a requirement for this module?

Comment (MikaelSmith) [inline manifests/install.pp]: Stringified facts yes, but the 3.8 requirement (on agents) is going away.

**Comment by highb** [comment]
Hopefully that's the last spec fix. Ever closer...

**Comment by highb** [comment]
@MikaelSmith Commits squashed, I preserved the rest of the commit messages in the squash commit for posterity. Windows specs updated to check against `1.2.1.1`.
